src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -56,7 +57,58 @@ private static IEnumerable<FunctionOverload> GetSystemOverloads(IFeatureProvider
                 .Build();
 
             yield return new FunctionOverloadBuilder("concat")
-                .WithReturnType(LanguageConstants.Array)
+                .WithReturnResultBuilder(TryDeriveLiteralReturnType("concat", (_, _, _, _, argumentTypes) =>
+                {
+                    if (argumentTypes.All(t => t is TupleType))
+                    {
+                        return new(new TupleType(argumentTypes.OfType<TupleType>().SelectMany(tt => tt.Items).ToImmutableArray(), default));
+                    }
+
+                    BigInteger minLength = 0;
+                    BigInteger? maxLength = null;
+                    var itemTypes = new ITypeReference[argumentTypes.Length];
+
+                    for (int i = 0; i < argumentTypes.Length; i++)
+                    {
+                        if (argumentTypes[i] is not ArrayType arr)
+                        {
+                            return new(LanguageConstants.Array);
+                        }
+
+                        itemTypes[i] = arr.Item;
+
+                        minLength += arr.MinLength ?? 0;
+
+                        if (i == 0)
+                        {
+                            maxLength = arr.MaxLength;
+                        }
+                        else if (maxLength.HasValue && arr.MaxLength.HasValue)
+                        {
+                            maxLength = maxLength.Value + arr.MaxLength.Value;
+                        }
+                        else
+                        {
+                            maxLength = null;
+                        }
+                    }
+
+                    return new(TypeFactory.CreateArrayType(TypeHelper.CreateTypeUnion(itemTypes),
+                        minLength switch
+                        {
+                            var zero when zero <= 0 => null,
+                            var tooBig when tooBig > long.MaxValue => long.MaxValue,
+                            var otherwise => (long) otherwise,
+                        },
+                        maxLength switch
+                        {
+                            null => null,
+                            var tooBig when tooBig > long.MaxValue => long.MaxValue,
+                            var otherwise => (long) otherwise,
```

Comment:
The bottom two branches can also be simplified to `var otherwise => (long)BigInteger.Min(otherwise, long.MaxValue),`